File: C/ROM.c
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <SDL2/SDL.h>
#include "./include/CHIP8.h"
#include "./include/Font.h"

void init_memory(struct CHIP8* self) {
    for (int i = 0x100; i < 0x150; i++) {
        self->memory[i] = HEX_SPRITES[i % 0x100];
    }
}

void read_rom(struct CHIP8* self, char* file_path) {

    FILE* rom = fopen(file_path, "rb");

    if (rom == NULL) {
        perror("Failed to open ROM file!\n");
        SDL_Quit();
    }

    //printf("----- ROM Read -----\n");
    int index = 0x200;

    while (true) {
        char byte; 
        fread(&byte, 1, 1, rom);

        if (feof(rom) != 0) {
            break;
        }

        self->memory[index] = (uint8_t) byte;
        //printf("[0x%x] = %x\n", index, self->memory[index]);
        index += 1;

        if (index >= 4096)
            break;
    }

    //printf("----- ROM Finish -----\n");
    fclose(rom);
}

File: Rust/src/rom.rs
use super::font;
use super::CHIP8;

pub fn init_memory(interpreter: &mut CHIP8) {
    for i in 0..16 {
        for j in 0..5 {
            interpreter.memory[(font::HEX_SPRITE_ADDRESS[i as usize] + j) as usize] = font::HEX_SPRITE[i as usize][j as usize];
        }
    }
}

pub fn read_rom(interpreter: &mut CHIP8) {
    let args = std::env::args().into_iter();
    //dbg!(&args);
    let file_path = args.last().unwrap();

    let rom = std::fs::read(file_path.clone()).unwrap();
    //println!("{}", file_path.clone());
    //println!("{}", rom.len());
    //std::process::exit(0);
    //let mut index: usize = 0x200;

    let limit = if (0x200 + rom.len()) < 4096 { 0x200 + rom.len() } else { 4096 };

    for i in 0x200..limit {
        //dbg!(rom[i - 0x200]);
        interpreter.memory[i] = rom[i - 0x200];
        //index += 1;
    }

    //std::process::exit(0);
}